### PR TITLE
V.2 Candidate

### PR DESCRIPTION
**Pull request by KarelWintersky**
* [+] More exceptions and error messages
* [+] Added ImageResizeInterface
* [*] Reorder and optimize code
* [*] GammaCorrection disabled by default (issue #162)
* [*] Reformat code
* [*] Tests: updated
* [*] Fixes: composer.json
* [*] Fixes: README.md

### CODE REVIEW

**Comment by KarelWintersky** [comment]
Two months... 

bump?

**Comment by peter279k** [comment]
Sorry. I fell of the stairs accidentally last Friday.

And I can't review your codes for now.

Once I recover, I will review your codes.

**Comment by KarelWintersky** [comment]
Well, I'll fork this and publish as our corporate-internal package.

**Comment by adityapatadia** [comment]
If you can wait till weekend, I can review and merge it.

File: composer.json [hunk 1]
```diff
@@ -1,45 +1,43 @@
 {
-    "name": "gumlet/php-image-resize",
```
Comment (peter279k) [inline composer.json:2]: I don't know why the space tab has been changed.

File: lib/ImageResize.php [hunk 2]
```diff
@@ -7,316 +7,333 @@
...
+     */
     public $interlace = 1;
-
+    
+    /**
+     * @var mixed
+     */
     public $source_type;
-
+
```
Comment (peter279k) [inline lib/ImageResize.php:46]: Additional space tabs.

File: lib/ImageResize.php [hunk 3]
```diff
@@ -7,316 +7,333 @@
...
     public static function createFromString($image_data)
     {
-        if (empty($image_data) || $image_data === null) {
-            throw new ImageResizeException('image_data must not be empty');
+        if (empty( $image_data ) || $image_data === null) {
+            throw new ImageResizeException( __CLASS__ . ' ERROR: image_data must not be empty.' );
         }
-        $resize = new self('data://application/octet-stream;base64,' . base64_encode($image_data));
-        return $resize;
+        return new self( 'data://application/octet-stream;base64,'.base64_encode( $image_data ) );
```
Comment (peter279k) [inline lib/ImageResize.php:73]: Please consider reverting the previous coding style.

File: lib/ImageResize.php [hunk 4]
```diff
@@ -7,316 +7,333 @@
...
-        if ($filename === null || empty($filename) || (substr($filename, 0, 5) !== 'data:' && !is_file($filename))) {
-            throw new ImageResizeException('File does not exist');
-        }
-
-        $finfo = finfo_open(FILEINFO_MIME_TYPE);
-        $checkWebp = false;
-        if (strstr(finfo_file($finfo, $filename), 'image') === false) {
-            if (version_compare(PHP_VERSION, '7.0.0', '<=') && strstr(file_get_contents($filename), 'WEBPVP8') !== false) {
-                $checkWebp = true;
+        if (!defined( 'IMAGETYPE_WEBP' )) {
```
Comment (peter279k) [inline lib/ImageResize.php:78]: Please consider reverting the previous coding style.

File: lib/ImageResizeInterface.php [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    /**
+     * Enable or not the gamma color correction on the image, enabled by default
+     *
+     * @param bool $enable
+     * @return static
+     */
+    public function gamma($enable = true);
+    
+    public function imageFlip($image, $mode);
+}
```
Comment (peter279k) [inline lib/ImageResizeInterface.php:218]: Missing the EOF, end of file.

File: lib/ImageResize.php [hunk 6]
```diff
@@ -712,78 +660,95 @@ protected function getCropPosition($expectedSize, $position = self::CROPCENTER)
...
+        return $this;
+    }
+    
+    public function setQualityWebp($quality)
+    {
+        $this->quality_webp = $quality ?: $this->quality_webp;
         return $this;
     }
+    
+
```
Comment (peter279k) [inline lib/ImageResize.php:753]: Additional new line.

**Review by peter279k — CHANGES_REQUESTED** [review]
Please consider following points:

- Reverting and be consistency for coding style.
- Checking the codes and remove useless new lines.

File: composer.json (hunk 1 above)
Comment (KarelWintersky) [reply composer.json:2]: is it a problem?

Comment (peter279k) [reply composer.json:2]: I think it should keep previous style.

File: lib/ImageResize.php (hunk 6 above)
Comment (KarelWintersky) [reply lib/ImageResize.php:753]: not an error

File: lib/ImageResize.php (hunk 4 above)
Comment (KarelWintersky) [reply lib/ImageResize.php:78]: code stylesheet required

File: lib/ImageResize.php (hunk 3 above)
Comment (KarelWintersky) [reply lib/ImageResize.php:73]: PHP compiler will optimize this code without any problems

File: composer.json (hunk 1 above)
Comment (KarelWintersky) [reply composer.json:2]: have no time for this

**Comment by KarelWintersky** [comment]
Please, give me linter configuration.

**Comment by KarelWintersky** [comment]
Please, add linter settings file

File: composer.json (hunk 1 above)
Comment (KarelWintersky) [reply composer.json:2]: because code is reformatted

Comment (peter279k) [reply composer.json:2]: Never mind. This is about the small problem and it can be fixed on another PR.

Comment (KarelWintersky) [reply composer.json:2]: блядь, не можешь докопаться до сути - доебись до офорграфии.

File: lib/ImageResize.php (hunk 3 above)
Comment (KarelWintersky) [reply lib/ImageResize.php:73]: I can't understand you. You need in `$resize` local value?

**Comment by ties-s** [comment]
Any news on this? The docs claim chaining works but if I save multiple sizes this creates really dark images

**Comment by KarelWintersky** [comment]
@ties-s , maintainers of gumlet/php-image-resize rejected any changes or pull requests. You can use my package: `ajur-media/php-image-resize`

**Comment by peter279k** [comment]
Sorry to my delay response.

Most of stuffs are good, but it looks like a breaking change.

It defines the Interface and some coding styles are not good.

If I've available time, I will try to fix some stuffs by myself.

**Comment by ties-s** [comment]
It’s about the gamma correction. While this may be a breaking change, it’s also a bug fix. The documentation claims chaining works fine, which it doesn’t right now. 
I think fixing this to align with docs and expected behavior should be a priority. You could always bump the version to indicate the breaking change.

**Comment by peter279k** [comment]
> 
> 
> It’s about the gamma correction. While this may be a breaking change, it’s also a bug fix. The documentation claims chaining works fine, which it doesn’t right now.
> I think fixing this to align with docs and expected behavior should be a priority. You could always bump the version to indicate the breaking change.

I know that, but I need to review them carefully by myself.

Anyway, I will create another PR to work on that.

**Comment by KarelWintersky** [comment]
> it’s also a bug fix. 

this bug fixed in my package ;-)

**Comment by adityapatadia** [comment]
@KarelWintersky this was closed by you and we did not reject this. It would be best if you can create smaller PRs which are easy to review and we can merge them quickly.

I really appreciate your contribution but moving too many things in one go can sometimes take more time to review. Hope you can help by creating smaller PRs.

**Comment by peter279k** [comment]
I agree with @adityapatadia because this PR is too huge to review. And most of code structured are changed a lot.

**Comment by KarelWintersky** [comment]
Now I have no time for this, sorry. 

Our project is under of emergency migration from MySQL to Postgresql. Deadline avg 26 Dec 2021.

**Comment by adityapatadia** [comment]
No problem. The gamma issue is fixed in latest version. The but of chaining not working will be resolved sometime soon.

**Comment by KarelWintersky** [comment]
You delayed the process of reviewing my code for so long, and then made so many ridiculous and meaningless demands (by refusing to provide the linter configuration or a document about YOUR codestyle), that I lost all desire to help you fix errors in YOUR library.

You couldn't find time for more than two months to review changes that clearly improved the performance of your code and fixed many problems.

After that you made a lot of ridiculous claims about my work: “extra lines”, “extra spaces”, and so on and so forth.

Sorry, but I don't feel like investing in your library anymore. I corrected all the mistakes that you made during its development and released my own based on your code according to the license you specified.

You can use my code in your projects.

**Pull request closed without merging**